src/fs_watcher.rs

```diff
@@ -0,0 +1,124 @@
+use crate::Dst;
+use crate::DstSender;
+use crate::EndpointMeta;
+use crate::Endpoints;
+use crate::Error;
+use inotify::{Event, EventMask, Inotify, WatchMask};
+use inotify_sys as ffi;
+use serde_json;
+use serde_yaml;
+use std::ffi::OsString;
+use std::mem;
+use std::string::String;
+use std::{fs, path::PathBuf};
+use tokio::stream::StreamExt;
+
+const EVENT_BUF_SZ: usize =
+    mem::size_of::<ffi::inotify_event>() + (libc::FILENAME_MAX as usize) + 1;
+
+#[derive(Debug)]
+pub struct FsWatcher {
+    endpoints_dir: PathBuf,
+    dst_sender: DstSender,
+}
+
+#[derive(Debug)]
+enum FileType {
+    Yaml,
+    Json,
+}
+
+#[derive(Debug)]
+pub struct FsWatcherError {
+    reason: String,
+}
+
+macro_rules! fs_watcher_error {
+    ($reason:expr) => {{
+        return Err(Box::new(FsWatcherError { reason: $reason }));
+    }};
+}
+
+// === impl FsWatcherError ===
+
+impl std::fmt::Display for FsWatcherError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&self.reason, f)
+    }
+}
+
+impl std::error::Error for FsWatcherError {}
+
+impl FsWatcher {
+    pub fn new(endpoints_dir: PathBuf, dst_sender: DstSender) -> Self {
+        Self {
+            endpoints_dir,
+            dst_sender,
+        }
+    }
+
+    fn parse_dst(file_name: &str) -> Result<(Dst, FileType), Error> {
+        let mut parts = file_name.rsplitn(2, ".");
+        match (parts.next(), parts.next()) {
+            (Some(ext), Some(name)) => {
+                let dst: Dst = name.parse()?;
+                let ft = match ext {
+                    "yaml" => FileType::Yaml,
+                    "json" => FileType::Json,
+                    _ => fs_watcher_error!(format!("invalid file ext {}", ext)),
+                };
+                Ok((dst, ft))
+            }
+            _ => fs_watcher_error!(format!("invalid file name {}", file_name)),
+        }
+    }
+
+    fn parse_file(&self, file_name: &str, ft: FileType) -> Result<Endpoints, Error> {
+        let path = self.endpoints_dir.join(file_name);
+        let contents = fs::read_to_string(path)?;
```

Comment:
Let's use [the version of `read_to_string` from `tokio::fs`](https://docs.rs/tokio/0.2.21/tokio/fs/fn.read_to_string.html) --- this is essentially an identical API, except it's an async function that spawns the filesystem work on Tokio's blocking threadpool so it doesn't block an async worker thread.

Note that this will require enabling Tokio's `"fs"` feature flag.